### PR TITLE
Rewrite draw_segmentation_masks and update gallery example to illustrate both instance and semantic segmentation models

### PR DESCRIPTION
**Pull request by NicolasHug**
This PR changes the API of `draw_segmentation_masks` to something more generic that can support all models. See https://github.com/pytorch/vision/pull/3820 for more details.

It also significantly updates the gallery example with didactic instructions about the models outputs, and how to draw the masks with the new API.

### CODE REVIEW

File: gallery/plot_visualization_utils.py [hunk 1]
```diff
@@ -99,19 +99,19 @@ def show(imgs):
...
-# transparency of masks.
-#
-# Here is demo with torchvision's FCN Resnet-50, loaded with
-# :func:`~torchvision.models.segmentation.fcn_resnet50`.
-# You can also try using
-# DeepLabv3 (:func:`~torchvision.models.segmentation.deeplabv3_resnet50`)
-# or lraspp mobilenet models
+# draw segmentation amasks on images.
+# 
+# Semantic segmentation models
```
Comment (NicolasHug) [inline gallery/plot_visualization_utils.py:104]: @datumbox To illustrate why natively handling floats images [0, 1] is useful see below how we never use the uint8 images. We only use the batch (floats in 0, 1) and the normalized batch.

The only reason we still need the uint8 images in this example is because `draw_bounding_boxes` doesn't support floats yet, but I intend to update it as well. Once done, we will be able to completely remove references to the  `uint8` images.

Comment (datumbox) [reply gallery/plot_visualization_utils.py:106]: If this PR requires the other one, consider merging the two or stacking them. Given the discussion on the other thread, it's worth showing end-to-end your proposal and that might lead to faster merging.

Comment (NicolasHug) [reply gallery/plot_visualization_utils.py:106]: This PR in already a superset of #3820 the changes of #3820 are included here, but I think they should be merged separately

File: gallery/plot_visualization_utils.py [hunk 2]
```diff
@@ -91,41 +90,264 @@ def show(imgs):
...
-from torchvision.utils import draw_segmentation_masks
 
 
 model = fcn_resnet50(pretrained=True, progress=False)
 model = model.eval()
 
-# The model expects the batch to be normalized
-batch = F.normalize(batch, mean=(0.485, 0.456, 0.406), std=(0.229, 0.224, 0.225))
-outputs = model(batch)
+normalized_batch = F.normalize(batch, mean=(0.485, 0.456, 0.406), std=(0.229, 0.224, 0.225))
```
Comment (datumbox) [inline gallery/plot_visualization_utils.py:124]: nit: I know from earlier discussions that you preferred to minimize the number of batch variables pointing to the image. If you still want to do this you can either assign back to batch or call `F.normalize()` with `inplace=True`. Up to you.

**Review by datumbox — APPROVED** [review]
It looks great to me. Thanks @NicolasHug.

I know that you were not convinced about removing the float support but thank you for removing it from the PR to let more discussions around the issue.

This specific enhancement looks very nice and I think it's a great idea to expand the tool to support semantic segmentation. For me this is good to merge.

File: gallery/plot_visualization_utils.py (hunk 2 above)
Comment (NicolasHug) [reply gallery/plot_visualization_utils.py:124]: I think I still need to keep the non-normalized `batch` variable, as it's needed by the Mask-RCNN model below.

File: gallery/plot_visualization_utils.py [hunk 3]
```diff
@@ -91,41 +90,264 @@ def show(imgs):
...
+
+sem_classes = [
+    '__background__', 'aeroplane', 'bicycle', 'bird', 'boat', 'bottle', 'bus',
+    'car', 'cat', 'chair', 'cow', 'diningtable', 'dog', 'horse', 'motorbike',
+    'person', 'pottedplant', 'sheep', 'sofa', 'train', 'tvmonitor'
+]
+sem_class_to_idx = {cls: idx for (idx, cls) in enumerate(sem_classes)}
+
+# We normalize the masks of each image in the batch independently
+normalized_masks = torch.stack([torch.nn.Softmax(dim=0)(masks) for masks in output])
```
Comment (fmassa) [inline gallery/plot_visualization_utils.py:146]: Can you instead do `torch.nn.functional.softmax(masks, dim=0)` or maybe `masks.softmax(dim=0)`?

Actually, you can probably just do
```python
normalized_masks = output.softmax(dim=1)
```
no?


I think it's an anti-pattern to instantiate the `nn.Module` just to perform this operation once

**Review by fmassa — APPROVED** [review]
Looks great, thanks a lot @NicolasHug !

I have a minor comment, otherwise looks good to me.

File: gallery/plot_visualization_utils.py (hunk 3 above)
Comment (NicolasHug) [reply gallery/plot_visualization_utils.py:146]: Ok, I'll use the functional API.

`normalized_masks = output.softmax(dim=1)` I think this would compute the softmax across the batch as well (dimension 0), and we wouldn't end up with values that sum to 1 for each image independently, the values would sum to 1 across the entire batch. That's actually something I wanted to double check with you, WDYT?

File: gallery/plot_visualization_utils.py (hunk 2 above)
Comment (datumbox) [reply gallery/plot_visualization_utils.py:124]: You are right, I did not see you reuse the same input for both models. Sounds good.

File: gallery/plot_visualization_utils.py (hunk 3 above)
Comment (fmassa) [reply gallery/plot_visualization_utils.py]: I think it should do the expected thing if we use `output.softmax(dim=1)`, but it's a good question.

We use it already in the loss for the segmentation models, which internally is [handled by passing `dim=1`](https://github.com/pytorch/pytorch/blob/4cb534f92ef6f5b2ec99109b0329f93a859ae831/aten/src/ATen/native/LossNLL.cpp#L442) to (log)softmax.

Internally, the softmax implementation splits the dimensions into those that are "batch dimensions" (which comes before the `dim`) and the "to be softmax" dimensions (which comes after `dim), see [here for more details](https://github.com/pytorch/pytorch/blob/4cb534f92ef6f5b2ec99109b0329f93a859ae831/aten/src/ATen/native/SoftMax.cpp#L19-L22)

**Comment by oke-aditya** [comment]
So glad that the implementation is improved now! :smiley: :tada:

**Comment by NicolasHug** [comment]
Thanks for your initial work @oke-aditya , it made this next version much easier to write and test!
